lib/AST/ArchetypeBuilder.cpp

```diff
@@ -587,13 +585,21 @@ auto ArchetypeBuilder::PotentialArchetype::getNestedType(
         if (!alias->hasInterfaceType())
           continue;
 
-        auto type = alias->getDeclaredInterfaceType();
-        if (auto existingPA = builder.resolveArchetype(type)) {
-          builder.addSameTypeRequirementBetweenArchetypes(pa, existingPA,
-                                                          redundantSource);
-        } else {
-          builder.addSameTypeRequirementToConcrete(pa, type, redundantSource);
-        }
+        // This interface type is in the context of the protocol, e.g. typealias
+        // X = Self.A.B.C, where Self is that of the protocol. The Self needs to
+        // be substituted for the current PA (which could be a dependent type
+        // like like T.X.Y), or else the .A.B.C look-up happens on whatever
+        // random type the τ_0_0 parameter happens to be.
+        auto typeInProtocol = alias->getDeclaredInterfaceType();
+        auto self = cast<GenericTypeParamType>(
+            proto->getSelfInterfaceType()->getCanonicalType());
+        auto concreteSelf = getDependentType({}, /*allowUnresolved=*/true);
+        SubstitutionMap subMap;
+        subMap.addSubstitution(self, concreteSelf);
+        subMap.addConformance(self, ProtocolConformanceRef(proto));
+        auto typeHere = typeInProtocol.subst(subMap);
```

Comment:
@slavapestov is going to be grouchy at building the substitution map this way. Can you just provide the trivial lookup functions that handle "self -> concreteSelf" and the construction of the abstract conformance? There's no need to build a full substitution map here.